src/marked.js

```diff
@@ -139,45 +139,99 @@ marked.defaults = defaults;
  */
 
 marked.use = function(extension) {
-  const opts = merge({}, extension);
-  if (extension.renderer) {
-    const renderer = marked.defaults.renderer || new Renderer();
-    for (const prop in extension.renderer) {
-      const prevRenderer = renderer[prop];
-      renderer[prop] = (...args) => {
-        let ret = extension.renderer[prop].apply(renderer, args);
-        if (ret === false) {
-          ret = prevRenderer.apply(renderer, args);
+  if (!Array.isArray(extension)) { // Wrap in array if not already to unify processing
+    extension = [extension];
+  }
+
+  const opts = merge({}, ...extension);
+  const extensions = marked.defaults.extensions || { renderers: {}, walkableTokens: {} };
+  let hasExtensions;
+
+  extension.forEach((pack) => {
+    // ==-- Parse "addon" extensions --== //
+    if (pack.extensions) {
+      hasExtensions = true;
+      pack.extensions.forEach((ext) => {
+        if (ext.renderer && ext.name) { // Renderers must have 'name' property
```

Comment:
I think all extensions should have a `name` property so it should probably throw an error if an extension doesn't.

Reply:
Also what happens if `name` already exists or is one of the original tokenizer/renderer names?

Reply:
I don't know. What behavior do we want?

Reply:
I think if it's the same as an existing one, the Tokenizer extension would work fine but the renderer would never trigger because the default renderer should consume any tokens with that name first.

Reply:
> the Tokenizer extension would work fine

I think currently it would run the extension tokenizer and the original if an extension contained the same name as a current tokenizer.

> It probably makes sense to allow the user to write an overwrite extension using the new structure so they don't have to juggle remembering multiple different styles.

I agree

Reply:
> I think currently it would run the extension tokenizer and the original

Yes.

> It probably makes sense to allow the user to write an overwrite extension using the new structure

In that case do we also need to require that they provide the correct `level: 'block' or 'inline`'? What's the preferred output if we get invalid syntax like that? Or if there's no `name` given on a normal extension? Do we give an error or just ignore the extension?

Reply:
Definitely throw an error. It will help with debugging. We should be as descriptive as we can. Like ` extension name required` and `extension level must be 'block' or 'inline'`

Reply:
Now throws errors for missing `name` on all extensions or invalid `level` on tokenizer extensions.